backend/svirt.pm

```diff
@@ -167,25 +170,33 @@ sub save_snapshot {
 
 sub load_snapshot {
     my ($self, $args) = @_;
-    my $snapname = $args->{name};
-    my $vmname   = $self->console('svirt')->name;
-    my $rsp;
+    my $snapname                   = $args->{name};
+    my $vmname                     = $self->console('svirt')->name;
     my $post_load_snapshot_command = '';
+    my $vm_state;
+    my $rsp;
     if (check_var('VIRSH_VMM_FAMILY', 'hyperv')) {
         my $ps = 'powershell -Command';
+        $self->run_ssh_cmd("$ps (Get-VMCheckpoint -VMName $vmname).Name");
+        (undef, $vm_state) = $self->run_ssh_cmd("$ps (Get-VM -VMName $vmname).State", wantarray => 1);
+        $vm_state =~ s/[\r\n]+$//;
         $rsp = $self->run_ssh_cmd(qq($ps "\$ProgressPreference='SilentlyContinue'; Restore-VMSnapshot -VMName $vmname -Name $snapname -Confirm:\$false"));
+        # after restoring from checkpoint, VM should be turned off.
+        bmwqemu::diag "VM state: $vm_state, after restoring from checkpoint";
         $self->run_ssh_cmd("mv -v xfreerdp_${vmname}_stop xfreerdp_${vmname}_stop.bkp", $self->get_ssh_credentials('hyperv'));
-
-        for my $i (1 .. 5) {
-            # Because of FreeRDP issue https://github.com/FreeRDP/FreeRDP/issues/3876,
-            # we can't connect too "early". Let's have a nap for a while.
-            sleep 10;
-            last
-              unless $self->run_ssh_cmd(
-                "pgrep --full --list-full xfreerdp.*\$(cat xfreerdp_${vmname}_stop.bkp)",
-                $self->get_ssh_credentials('hyperv'));
-            $self->die("xfreerdp did not start") if ($i eq 5);
-        }
+        # Because of FreeRDP issue https://github.com/FreeRDP/FreeRDP/issues/3876,
+        # we can't connect too "early". Let's have a nap for a while.
+        my $i = 0;
+        do {
```

Comment:
Not sure about this loop. When I take a look to the log it looks like:
```
2021-05-26T23:03:32.508 CEST] [debug] VM state: Running, after restoring from checkpoint
....
2021-05-26T23:03:38.918 CEST] [debug] VM state: Running, attempts: 1
```
So we restore the VM and get state `Running` but still checking for `Running`.
Maybe we need to check something different, but from my rdp experiences it isn't that easy. 
Before we had that `pgrep xfreerdp` in the loop. Now you checking for state and reduce the minimum sleep from 10s to 5s.

Somehow I'm not sure to understand the behavior here from code reading, can you throw some light on it?

Reply:
And who is staring this `freerdp.*$(cat xfreerdp_${vmname}_stop.bkp)` ?

Reply:
Yeah, good point, I have changed it because of _Production_ checkpoint was enabled instead of _Standard_ therefore the VM triggered boot point. I haven't touched this part afterwards and I agree it should be changed.

Reply:
Is it this loop: https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/tests/installation/bootloader_hyperv.pm#L244 ?

So actually the VM state doesn't have any meaning, or? But we need to check that `xfreerdp` is running for appropriate time or maybe better we could search in `$xfreerdp_log` for some success message?!

Reply:
> Is it this loop: https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/tests/installation/bootloader_hyperv.pm#L244 ?
> 

Not needed, it should reconnect by itself, but it takes time, so we cannot proceed further with testing as there is only black screen.

> So actually the VM state doesn't have any meaning, or? But we need to check that `xfreerdp` is running for appropriate time or maybe better we could search in `$xfreerdp_log` for some success message?!

It can help us to determine wrong checkpoint type settings. In case of default settings (Production), the VM would be in state _Off_ and not in _Running_.

Reply:
[Production snapshot test](http://kepler.suse.cz/tests/5129#) where the VM is turned off after restoring from checkpoint.

```
[0m[37m[2021-05-27T13:37:18.155 CEST] [debug] [run_ssh_cmd(powershell -Command (Get-VM -VMName openQA-SUT-8).State)] exit-code: 0
[0m[37m[2021-05-27T13:37:18.156 CEST] [debug] Restoring RDP connection, attempts: 1, VM state: Off
[0m[2021-05-27T13:37:18.156 CEST] [debug] <<< backend::baseclass::run_ssh_cmd(cmd="pgrep --full --list-full xfreerdp.*\$(cat xfreerdp_openQA-SUT-8_stop.bkp)", wantarray=0, hostname="openqaw5-xen-1.qa.suse.de", username="root", password="SECRET", keep_open=1)
[2021-05-27T13:37:18.156 CEST] [debug] <<< backend::baseclass::run_ssh(cmd="pgrep --full --list-full xfreerdp.*\$(cat xfreerdp_openQA-SUT-8_stop.bkp)", username="root", hostname="openqaw5-xen-1.qa.suse.de", wantarray=0, keep_open=1, password="SECRET")
[2021-05-27T13:37:18.157 CEST] [debug] <<< backend::baseclass::new_ssh_connection(username="root", blocking=1, wantarray=0, password="SECRET", hostname="openqaw5-xen-1.qa.suse.de", keep_open=1)
[37m[2021-05-27T13:37:18.212 CEST] [debug] Use existing SSH connection (key:hostname=openqaw5-xen-1.qa.suse.de,username=root)
[0m[37m[2021-05-27T13:37:18.276 CEST] [debug] [run_ssh_cmd(pgrep --full --list-full xfreerdp.*$(cat xfreerdp_openQA-SUT-8_stop.bkp))] exit-code: 1
[0m[33m[2021-05-27T13:37:18.276 CEST] [info] ::: backend::baseclass::die_handler: Backend process died, backend errors are reported below in the following lines:
  The VM is not running at /usr/lib/os-autoinst/backend/svirt.pm line 337.
  ```